mobile/src/hoc/withWeb3Accounts.js

```diff
@@ -0,0 +1,279 @@
+        DeviceEventEmitter.emit('messagingKeys', messagingKeys)
+      }
+    }
+
+    /* Create new account
+     */
+    createAccount = () => {
+      const mnemonic = ethers.Wallet.createRandom().mnemonic
+      // This is the default path but explicitly stated here for clarity
+      const derivePath = `m/44'/60'/0'/0/0`
+      // Web3js doesn't support wallet creation from a mnemonic, so somewhat
+      // redundantly we have to include ethersjs. Perhaps migrate everything
+      // away from web3js to ethersjs or the functionality will be added to web3js
+      // sometime in the future, see:
+      // https://github.com/ethereum/web3.js/issues/1594
+      const account = ethers.Wallet.fromMnemonic(mnemonic, derivePath)
+      this.addAccount({
+        address: account.address,
+        mnemonic: account.mnemonic,
+        privateKey: account.privateKey
+      })
+    }
+
+    /* Add a new account based on a mnemonic. If the first account
+     * in the derivation path is used it will continue to try the next number
+     * until an address that is unused is found.
+     *
+     * This function can be slow.
+     */
+    importAccountFromMnemonic = (mnemonic, maxRetries = 10) => {
+      const existingAddresses = this.props.wallet.accounts.map(a => a.address)
+      // Use a loop to try the next account in the derivation path
+      for (let i = 0; i < maxRetries; i++) {
+        // This is the default path but explicitly stated here for clarity
+        const derivePath = `m/44'/60'/0'/0/${i}`
+        // Web3js doesn't support wallet creation from a mnemonic, so somewhat
+        // redundantly we have to include ethersjs. Perhaps migrate everything
+        // away from web3js to ethersjs or the functionality will be added to web3js
+        // sometime in the future, see:
+        // https://github.com/ethereum/web3.js/issues/1594
+        const account = ethers.Wallet.fromMnemonic(mnemonic, derivePath)
+        if (!existingAddresses.includes(account.address)) {
+          // Got an account we don't have, use that
+          this.addAccount({
+            address: account.address,
+            mnemonic: account.mnemonic,
+            privateKey: account.privateKey
+          })
+          return account
+        }
+      }
+      throw new Error('Maximum addresses reached')
+    }
+
+    /* Add a new account based on a private key (this.state.value).
+     */
+    importAccountFromPrivateKey = privateKey => {
+      if (!privateKey.startsWith('0x') && /^[0-9a-fA-F]+$/.test(privateKey)) {
+        privateKey = '0x' + privateKey
+      }
+      const account = new ethers.Wallet(privateKey)
+      this.addAccount({
+        address: account.address,
+        mnemonic: account.mnemonic,
+        privateKey: account.privateKey
+      })
+      return account
+    }
+
+    /* Add a new account
+     */
+    addAccount = account => {
+      this.props.addAccount(account)
+      global.web3.eth.accounts.wallet.add(account)
```

Comment:
should this kind of thing be happening in the action? Not a big deal but think business logic should generally live in there

Reply:
Thanks, yea, there should probably be some simplifying here. It followed a bit of a weird evolution to get here.

Reply:
Moved a lot of the logic into redux and out of this HOC